File: src/main/java/infinitespire/powers/RealityShiftPower.java
package infinitespire.powers;

import com.megacrit.cardcrawl.cards.DamageInfo;
import com.megacrit.cardcrawl.core.CardCrawlGame;
import com.megacrit.cardcrawl.dungeons.AbstractDungeon;
import com.megacrit.cardcrawl.localization.PowerStrings;
import com.megacrit.cardcrawl.powers.AbstractPower;
import infinitespire.InfiniteSpire;
import infinitespire.monsters.Nightmare;

public class RealityShiftPower extends AbstractPower {
	public Nightmare nightmare;
	public boolean hasTriggered = false;

	public static final String powerID = InfiniteSpire.createID("RealityShiftPower");
	private static final PowerStrings strings = CardCrawlGame.languagePack.getPowerStrings(powerID);
	public final int maxAmount;

	public RealityShiftPower(Nightmare nightmare, int amount) {
		this.owner = nightmare;
		this.nightmare = nightmare;
		this.amount = amount;
		this.maxAmount = amount;
		this.name = strings.NAME;
		this.ID = powerID;
		this.img = InfiniteSpire.Textures.getPowerTexture("realityshift.png");
		this.type = PowerType.BUFF;
		this.isTurnBased = true;
		this.priority = -99999;
		this.updateDescription();
	}
	
	
	
	@Override
	public void atEndOfTurn(boolean isPlayer) {
		if(!isPlayer){
			this.hasTriggered = false;
			this.amount = this.maxAmount;
			this.updateDescription();
		}
	}

	@Override
	public int onAttacked(DamageInfo info, int damageAmount) {
		this.amount -= damageAmount;
		if(this.amount <= 0 && !hasTriggered){
			nightmare.onEnoughDamageTaken();
			this.amount = 0;
			this.hasTriggered = true;
		}
		return super.onAttacked(info, damageAmount);
	}

	public int getAttackStr() {
		DamageInfo dmg = nightmare.damage.get(0);
		dmg.applyPowers(nightmare, AbstractDungeon.player);

		return dmg.output;
	}

	public void updateDescription() {
		this.description =
			strings.DESCRIPTIONS[0]
				+ amount
				+ strings.DESCRIPTIONS[1]
				+ getAttackStr()
				+ strings.DESCRIPTIONS[2]
				+ (nightmare.effectCount + 1)
				+ strings.DESCRIPTIONS[3];
	}
}

File: src/main/kotlin/com/blanktheevil/infinitespire/powers/RealityShiftPower.kt
package com.blanktheevil.infinitespire.powers

import basemod.interfaces.CloneablePowerInterface
import com.blanktheevil.infinitespire.extensions.languagePack
import com.blanktheevil.infinitespire.extensions.makeID
import com.blanktheevil.infinitespire.extensions.player
import com.blanktheevil.infinitespire.monsters.Nightmare
import com.blanktheevil.infinitespire.powers.utils.PowerBuilder
import com.megacrit.cardcrawl.cards.DamageInfo
import com.megacrit.cardcrawl.powers.AbstractPower

class RealityShiftPower(val nightmare: Nightmare, amount: Int) :
    AbstractPower(),
    CloneablePowerInterface {
  companion object {
    val powerID = "RealityShiftPower".makeID()
    val BUILDER = PowerBuilder(powerID)
      .img("realityshift.png")
      .isTurnBased()
      .buff()
      .priority(Int.MIN_VALUE)
  }

  val strings = languagePack.getPowerStrings(powerID)

  init {
    this.owner = nightmare
    this.amount = amount
    this.name = strings.NAME
    this.ID = powerID
    this.img = BUILDER.img
    this.type = BUILDER.type
    this.priority = BUILDER.priority
    updateDescription()
  }

  private val maxAmount = amount
  private var hasTriggered = false

  override fun atEndOfTurn(isPlayer: Boolean) {
    if (!isPlayer) {
      hasTriggered = false
      this.amount = maxAmount
      this.updateDescription()
    }
  }

  override fun onAttacked(info: DamageInfo, damageAmount: Int): Int {
    this.amount -= damageAmount
    if (this.amount <= 0 && !hasTriggered) {
      this.amount = 0
      nightmare.triggerRealityShiftAttack()
      this.hasTriggered = true
    }

    return super.onAttacked(info, damageAmount)
  }

  private fun getAttackStr(): Int {
    val dmg = nightmare.damage[0]
    dmg.applyPowers(nightmare, player)
    return dmg.output
  }

  override fun updateDescription() {
    with(strings) {
      description = StringBuilder()
        .append(DESCRIPTIONS[0])
        .append(amount.toString())
        .append(DESCRIPTIONS[1])
        .append(getAttackStr())
        .append(DESCRIPTIONS[2])
        .append(nightmare.attackAmount + 1)
        .append(DESCRIPTIONS[3])
        .toString()
    }
  }

  override fun makeCopy(): AbstractPower {
    return RealityShiftPower(this.owner as Nightmare, this.maxAmount)
  }
}